### PR TITLE
Align behavior of pre-commit hook and Sphinx extension

### LINKED ISSUE #466: Align pre-commit and Sphinx rule override syntax
See https://github.com/numpy/numpydoc/pull/454

/cc @stefmolin

### PR DESCRIPTION
**Pull request by stefmolin**
Closes #466 

Changes made:
- Extracted validation check selection logic from the Sphinx extension into a utility function and added test cases.
- Changed the pre-commit hook check selection from an ignore list introduced in #454 to the same strategy used by `numpydoc_validation_checks` on the Sphinx side.
- Added global exclusion option to pre-commit hook mirroring `numpydoc_validation_exclude`.
- Deprecated the `override_GL08` pre-commit option, which is covered by the global exclusion option.
- Expanded the override logic to work for any check.
- Ported inline ignore comments logic from pre-commit side over to Sphinx extension and `--validate` sides.
- Brought override logic over to Sphinx extension side by creating `numpydoc_validation_overrides`, which allows the same flexibility that the `override_<code>` options in the pre-commit hook provide. Included this option in the `doc/install.rst` page and added some test cases.
- Updated pre-commit docs for changes in behavior.
- Created separate section for inline ignore comments in the docs since their scope expanded.
- Added cross-references between validation options in `doc/validation.rst`.

### CODE REVIEW

**Comment by stefmolin** [comment]
I think I have an issue with the type hints I added &ndash; I will comment when this is ready for review since I can't mark it as a draft.

**Comment by stefmolin** [comment]
Issue has been solved. Ready for review.

File: doc/validation.rst [hunk 1]
```diff
@@ -22,44 +24,55 @@ command line options for this hook:
...
         "SA01",
         "ES01",
     ]
-    override_SS05 = '^((Process|Assess|Access) )'
-    override_GL08 = '^(__init__)$'
+    exclude = [  # don't report on objects that match any of these regex
+        '\.undocumented_method$',
+        '\.__repr__$',
+    ]
+    override_SS05 = [  # override SS05 to allow docstrings starting with these words
```
Comment (stefanv) [inline doc/validation.rst:59]: Should this be a mapping, with the rule as the key?

File: doc/validation.rst [hunk 2]
```diff
@@ -22,44 +24,55 @@ command line options for this hook:
...
+  the form of ``override_<code>`` with a collection of regular expression(s)
+  to search for in the contents of a docstring, not the object name. This
+  maps to ``numpydoc_validation_overrides`` from the
+  :ref:`Sphinx build configuration <validation_during_sphinx_build>`.
 
 ``pyproject.toml``::
 
     [tool.numpydoc_validation]
-    ignore = [
+    checks = [
```
Comment (stefanv) [inline doc/validation.rst:49]: I've always found the numpydoc convention confusing; I think the exclusion list you had before is more logical.

**Review by stefanv — COMMENTED** [review]
@stefmolin Thanks for this refactor, it looks great!

I should have written out #466 more clearly, because I actually liked the way you had it implemented; it felt much more similar to how other tools like `ruff` select their rules.

Since you have now looked at this code quite a bit, I'd like to hear your opinion on how it should look, in an ideal world.

I still think we can align the two, but am hoping we can deprecate the  "Sphinx way" for the "pre-commit way".

@rossbar I think you've tried to explain to me before why numpydoc is a bit weird in the way it specifies rules, so perhaps you also want to weigh in.

File: doc/validation.rst (hunk 1 above)
Comment (stefmolin) [reply doc/validation.rst:59]: Nope &ndash; that format is only needed for the Sphinx side in order to find all of the related options. Here you just create values for each override you want e.g., `override_SS05` and `override_EX01`.

**Comment by stefmolin** [comment]
@stefanv - I probably should have double-checked with you as well. Personally, I have not used the Sphinx extension for validation, so I can't speak to that. For my original use case, I wrote the pre-commit hook we merged last week and set it up on my personal and work projects to ignore certain rules. I imagine that the rule set is unlikely to change much if at all in the future and that most use cases for the Sphinx side are not just specifying a list of rules to run, but rather to run all and ignore a select few; therefore, I would agree with you that an ignore list as I had initially with the pre-commit hook is much easier to work with and more logical. The only reason I mirrored the Sphinx extension approach was to avoid breaking other projects' documentation builds. If we are comfortable with that though, then I am more than happy to switch both to the ignore list approach.

**Comment by stefanv** [comment]
What do you think, @rossbar?

**Comment by rossbar** [comment]
I have no preference re: how the configuration is implemented (i.e. an "allowlist", an "ignorelist", both, etc.)

FWIW [the sphinx configuration](https://numpydoc.readthedocs.io/en/latest/install.html) supports both; i.e. either specifying a subset as an `allowlist`, or a "run-everything-except-these" `ignorelist`.

> I think you've tried to explain to me before why numpydoc is a bit weird in the way it specifies rules, so perhaps you also want to weigh in.

The thing that I don't like about the way the validation is currently structured is the validator runs *all* the checks no matter what, and the configuration only affects which are reported. IMO it'd be a major improvement to refactor the validation so that it is modular, and only the selected checks are actually **run**. This is way out of scope for this discussion though!

> I imagine that the rule set is unlikely to change much if at all in the future

I'd tend to agree - in fact, I think many of the implemented rules are already overly-specific to the point of not being very helpful.

> and that most use cases for the Sphinx side are not just specifying a list of rules to run, but rather to run all and ignore a select few.

IME in trying this out for large, existing doc sets (e.g. numpy, scipy, networkx, etc.) - this hasn't been the case. Most of the validation checks fire way too often and add a bunch of noise, to the point where there are only a few that are worth enabling. Automatic reformatting (a la linters) would help alleviate this a lot, but again that's for another feature discussion.

**Comment by stefmolin** [comment]
> FWIW [the sphinx configuration](https://numpydoc.readthedocs.io/en/latest/install.html) supports both; i.e. either specifying a subset as an `allowlist`, or a "run-everything-except-these" `ignorelist`.

With the changes currently in this PR, the pre-commit hook also supports both using the same logic.

> IME in trying this out for large, existing doc sets (e.g. numpy, scipy, networkx, etc.) - this hasn't been the case. Most of the validation checks fire way too often and add a bunch of noise, to the point where there are only a few that are worth enabling. Automatic reformatting (a la linters) would help alleviate this a lot, but again that's for another feature discussion.

In that case, perhaps it makes sense to use the Sphinx configuration for the pre-commit hook since you can provide a few rules to use rather than having to specify many to ignore.

**Comment by jarrodmillman** [comment]
@larsoner @rossbar @stefanv I would like to get this merged and then make a v1.6.0rc1 release later today. There seems to be agreement that this is an improvement and we can improve things in a future release if we want to.

Thoughts?

**Comment by larsoner** [comment]
+1 from me! You'll need to override branch protection rules to merge since pre-commit.ci won't run on this PR

**Review by jarrodmillman — APPROVED** [review]

**Comment by stefanv** [comment]
Let's get this in, and work towards refactoring the checking to not do unnecessary work in the future. Thanks so much @stefmolin!

**Comment by larsoner** [comment]
@jarrodmillman looks like this broke MNE-Python:

https://app.circleci.com/pipelines/github/mne-tools/mne-python/20169/workflows/d927bab9-9954-49ce-9cb2-2298f58cd29b/jobs/57313

Probably should sort this out before releasing!

File: numpydoc/validate.py [hunk 3]
```diff
@@ -506,9 +578,16 @@ def validate(obj_name, validator_cls=None, **validator_kwargs):
     else:
         doc = validator_cls(obj_name=obj_name, **validator_kwargs)
 
+    # lineno is only 0 if we have a module docstring in the file and we are
+    # validating that, so we change to 1 for readability of the output
+    ignore_validation_comments = extract_ignore_validation_comments(
+        doc.source_file_name
+    ).get(doc.source_file_def_line or 1, [])
```
Comment (larsoner) [inline numpydoc/validate.py:585]: ... I think this needs a `if doc.source_file_name is not None` perhaps?

Comment (stefanv) [reply numpydoc/validate.py:585]: Thanks for the heads-up!

Comment (stefmolin) [reply numpydoc/validate.py:585]: How did the Sphinx validation behave before when the `source_file_name` is `None`? Do you still run the validation?

Comment (larsoner) [reply numpydoc/validate.py:585]: Yes. We're generate some doc strings on the fly, I assume that's what's breaking this but not sure

**Pull request merged**